### PR TITLE
[Quant][Feature] Support online MXFP8 quantization for MoE and dense models

### PR DESCRIPTION
**Pull request by EdalatiAli**
## Purpose
Add support for **online MXFP8 quantization** (`--quantization mxfp8`), enabling BF16/FP16 models to be dynamically quantized to MXFP8 (microscaling FP8 with block-32 scales) at load time — for both **linear layers** and **MoE expert layers**.

This is powered by the FlashInfer kernels
- `trtllm_fp8_block_scale_moe` for MoE layers (see  #35986)
- `mm_mxfp8` for linear layers (see #35053 )

This PR implements part of the online quantization support proposed in #32029  and #32412 

### Usage

```bash
# Serve any BF16 model with online MXFP8 quantization
vllm serve <model> --quantization mxfp8

# Or with layer skipping via config.json:
# "quantization_config": {
#     "quant_method": "mxfp8",
#     "modules_to_not_convert": ["lm_head", "mlp.gate"]
# }
```

Requires **SM 100+** (Blackwell) GPU.

## Test Plan

E2E tests for both a dense model (`Qwen/Qwen3-0.6B`) and a MoE model (`Qwen/Qwen3-30B-A3B`): logprobs comparison against BF16 baseline + generation smoke test.

```bash
python -m pytest tests/models/quantization/test_mxfp8.py -v
```

In addition, we report the the accuracy on MMLU_pro and GM8K using lm_eval_harness as well as performance benchmarks for `Qwen/Qwen3-30B-A3B`. 

### Accuracy
```bash
lm_eval \
  --model vllm \
  --trust_remote_code \
  --model_args pretrained=Qwen/Qwen3-30B-A3B,quantization=mxfp8,enforce_eager=True \
  --tasks gsm8k,mmlu_pro \
  --batch_size auto
  ```
  
| Task | BF16 | MXFP8 |
|-----------|------|-------|
| GSM8K  (flexible-extract)   |   85.7   |   87.8    |
| GSM8K  (strict-match)   |   89.7   |   88.2    |
| MMLU-Pro  |   69.2   |    68.9   |

### Performance
```bash
vllm bench throughput --model Qwen/Qwen3-30B-A3B \
--tensor-parallel-size 1 \
--trust-remote-code \
--async-scheduling \
--backend vllm \
--dataset-name random \
--random-prefix-len 0 \
--random-input-len 1024 \
--random-output-len 1024 \
--max-num-seqs 128 \
--num-prompts 512 \
--quantization mxfp8    ## Remove for the bf16 model
```

#### BF16 performance
`Throughput: 7.73 requests/s, 15833.85 total tokens/s, 7916.92 output tokens/s`

#### MXFP8 performance
`Throughput: 10.34 requests/s, 21179.37 total tokens/s, 10589.68 output tokens/s`

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/fp8.py [hunk 1]
```diff
@@ -1297,6 +1365,157 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
+                intermediate_size_per_partition // MXFP8_BLOCK_SIZE,
+                dtype=torch.uint8,
+            ),
+            requires_grad=False,
+        )
+        layer.register_parameter("w13_weight_scale", w13_weight_scale)
+        layer.register_parameter("w2_weight_scale", w2_weight_scale)
+        set_weight_attrs(w13_weight_scale, extra_weight_attrs)
+        set_weight_attrs(w2_weight_scale, extra_weight_attrs)
+        layer.weight_block_size = [1, 32]
```
[the remark this answers is not in the document]
Comment (EdalatiAli) [reply vllm/model_executor/layers/quantization/fp8.py]: This issue is resolved by renaming the scale name postfix to `weight_scale`

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py [hunk 2]
```diff
@@ -212,6 +218,64 @@ def _apply_per_block(
...
+        num_expert_group: int | None = None,
+        e_score_correction_bias: torch.Tensor | None = None,
+        routed_scaling_factor: float | None = None,
+        topk_group: int | None = None,
+    ) -> torch.Tensor:
+        import flashinfer
+        from flashinfer.fused_moe import Fp8QuantizationType
+
+        assert not apply_router_weight_on_input
+        assert activation == MoEActivation.SILU
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:360]: It seems like you assert SILU but don't restrict selection in `_supports_activation`, is this necessary?

File: vllm/model_executor/layers/fused_moe/oracle/fp8.py [hunk 3]
```diff
@@ -221,6 +223,18 @@ def select_fp8_moe_backend(
...
+    if (
+        current_platform.is_cuda()
+        and current_platform.is_device_capability_family(100)
+        and weight_key == kMxfp8Static
+        and activation_key == kMxfp8Dynamic
+        and Fp8MoeBackend.FLASHINFER_TRTLLM in AVAILABLE_BACKENDS
+    ):
+        AVAILABLE_BACKENDS.remove(Fp8MoeBackend.FLASHINFER_TRTLLM)
+        AVAILABLE_BACKENDS.insert(0, Fp8MoeBackend.FLASHINFER_TRTLLM)
+
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/oracle/fp8.py]: We already have an mxfp8 oracle at https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/oracle/mxfp8.py, could we use that rather than overloading fp8?

File: vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py (hunk 2 above)
Comment (EdalatiAli) [reply vllm/model_executor/layers/fused_moe/experts/trtllm_fp8_moe.py:360]: Yes, it's necessary — the monolithic class serves both block-scale (SILU only, hardcoded in flashinfer) and per-tensor (SILU + RELU2) paths, and _supports_activation can't distinguish them. Adding quant context to _supports_activation would require changing the abstract interface and every implementation across the codebase.

File: vllm/model_executor/layers/fused_moe/oracle/fp8.py (hunk 3 above)
Comment (EdalatiAli) [reply vllm/model_executor/layers/fused_moe/oracle/fp8.py]: Sure, I changed the code to use [https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/oracle/mxfp8.py](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/fused_moe/oracle/mxfp8.py). I had to make a few other minor changes to correctly use it.

File: vllm/model_executor/layers/fused_moe/oracle/mxfp8.py [hunk 4]
```diff
@@ -1,44 +1,87 @@
...
+    {
+        Fp8MoeBackend.FLASHINFER_TRTLLM,
+    }
+)
 
-class MxFp8MoeBackend(Enum):
-    FLASHINFER_TRTLLM = "FLASHINFER_TRTLLM"
+_BACKEND_NAME_MAP: dict[str, Fp8MoeBackend] = {
+    "flashinfer_trtllm": Fp8MoeBackend.FLASHINFER_TRTLLM,
+}
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/oracle/mxfp8.py:26]: It is a bit confusing to use `Fp8MoeBackend` here and elsewhere for mxfp8, but I guess it is needed to reuse the moe utils

**Review by mgoin — APPROVED** [review]
Okay seems reasonable to me to accept. I'm not sure how much we should truly reuse with the fp8 methods but it is fair enough to follow that pattern for now

File: vllm/model_executor/layers/fused_moe/oracle/mxfp8.py (hunk 4 above)
Comment (EdalatiAli) [reply vllm/model_executor/layers/fused_moe/oracle/mxfp8.py:26]: Yes, using `Fp8MoeBackend` was needed to reuse the rest of the FP8 moe utils. 
We can follow a better approach when more MXFP8 backend are available in the future.
Thank you for the feedback!
